### PR TITLE
Add Wasm SjLj support

### PR DESCRIPTION
**Pull request by aheejin**
This adds SjLj handling support using Wasm EH instructions:
https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md

This does not pass CI currently; it will pass after
https://reviews.llvm.org/D108960 lands. This does not yet support mixing
of EH and SjLj.

These are SjLj tests that I couldn't attach `@with_both_sjlj_handling`
and the reasons:
- `test_dlfcn_longjmp`: Uses NodeRAWFS
- `test_longjmp_throw`: Mixes EH and SjLj
- `test_exceptions_longjmp1`: Mixes EH and SjLj
- `test_exceptions_longjmp2`: Mixes EH and SjLj
- `test_exceptions_longjmp3`: Mixes EH and SjLj
- `test_longjmp`: This currently uses `@also_with_standalone_wasm`, and
  it looks a test can't take two decorators when both of them take
  arguments, which is the case for `@also_with_standalone_wasm` and
  `@with_both_sjlj_handling`. Am I missing something?

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1342,6 +1342,16 @@ def default_setting(name, new_default):
   if settings.DISABLE_EXCEPTION_THROWING and not settings.DISABLE_EXCEPTION_CATCHING:
     exit_with_error("DISABLE_EXCEPTION_THROWING was set (probably from -fno-exceptions) but is not compatible with enabling exception catching (DISABLE_EXCEPTION_CATCHING=0). If you don't want exceptions, set DISABLE_EXCEPTION_CATCHING to 1; if you do want exceptions, don't link with -fno-exceptions")
 
+  # We cannot enable Emscripten SjLj and Wasm SjLj at the same time
+  if settings.SUPPORT_LONGJMP and settings.SJLJ_HANDLING:
```
Comment (sbc100) [inline emcc.py:1346]: `SUPPORT_LONGJMP` is enabled by default in `settings.py`.     Perhaps the best thing to do would be to disabled it by default when `SJLJ_HANDLING` is enaled.

I think the way to do this would be to add these two lines just before this check:

```
if settings.SJLJ_HANDLINGL
  default_settings('SUPPORT_LONGJMP', 0)
```

On the other hand if you do that it might confuse code like this in library.js:
https://github.com/emscripten-core/emscripten/blob/f099e309d6ab1b2ec6e1967a72d5e4e90ea19abf/src/library.js#L1498

Would it make sense to instead say that `SUPPORT_LONGJMP` means that longjmp support is enabled in one way or another (i.e. SJLJ_HANDLING implies SUPPORT_LONGJMP)?

File: src/settings.js [hunk 2]
```diff
@@ -1761,6 +1761,9 @@ var USES_DYNAMIC_ALLOC = 1;
 // library support.
 var SUPPORT_LONGJMP = 1;
 
+// WebAssembly setjmp-longjmp handling using Wasm EH (experimental)
+var SJLJ_HANDLING = 0;
```
Comment (sbc100) [inline src/settings.js:1765]: How about using `SUPPORT_LONGJMP=wasm` which would avoid a new setting and trying to keep them in sync?

File: tools/system_libs.py [hunk 3]
```diff
@@ -589,6 +589,20 @@ class Exceptions(IntEnum):
...
+  This represents setjmp/longjmp handling mode of Emscripten. Currently there are
+  two modes of exception handling:
+  - Emscripten: Emscripten provides setjmp/longjmp handling capability using JS
+    emulation. This causes code size increase and performance degradation.
+  - Wasm: Wasm setjmp/longjmp handling support uses Wasm EH instructions and
+    is meant to be fast. You need to use a VM that has the EH support to use
+    this. This is not fully working yet and still experimental.
+  """
+  EMSCRIPTEN = auto()
+  WASM = auto()
```
Comment (sbc100) [inline tools/system_libs.py:603]: Maybe we can avoid this extra enum too if we use `SUPPORT_LONGJMP=wasm` ?

File: tests/test_core.py [hunk 4]
```diff
@@ -923,31 +949,38 @@ def test_regex(self):
...
+  @with_both_sjlj_handling
   def test_longjmp2(self):
     self.do_core_test('test_longjmp2.c')
 
   @needs_dylink
+  @with_both_sjlj_handling
   def test_longjmp2_main_module(self):
     # Test for binaryen regression:
     # https://github.com/WebAssembly/binaryen/issues/2180
-    self.set_setting('MAIN_MODULE')
```
Comment (kripken) [inline tests/test_core.py:933]: Was this removal intentional?

File: tools/building.py [hunk 5]
```diff
@@ -458,9 +462,13 @@ def link_lld(args, target, external_symbols=None):
   # wasm-ld needs to take this for the LTO mode.
   if settings.EXCEPTION_HANDLING:
     cmd += ['-mllvm', '-exception-model=wasm', '-mllvm', '-wasm-enable-eh']
+  # Wasm SjLj handling. This is a CodeGen option for the LLVM backend, so
+  # wasm-ld needs to take this for the LTO mode.
+  if settings.EXCEPTION_HANDLING or settings.SJLJ_HANDLING:
+    cmd += ['-mllvm', '-exception-model=wasm']
```
Comment (kripken) [inline tools/building.py:468]: It's nice that lld has a single flag for both wasm exceptions and wasm sjlj. I wonder if we can do the same in emscripten somehow, and avoid adding a new flag here? But I guess we can't because we want the possibility to use wasm sjlj without C++ exceptions, since if we enable exceptions we change codegen and which system libraries are used. I'm curious how lld avoids that - I guess since it is not in charge of C++ semantics, and doesn't pick system libraries?

File: src/settings.js (hunk 2 above)
Comment (kripken) [reply src/settings.js:1765]: We could add some code to turn `-s SUPPORT_LONGJMP=1` to write `emscripten`, perhaps?

I am personally a little unsure what is best here. With exceptions we do have two separate options - we didn't try to have a single one for the 3 modes (none, emscripten, wasm). So maybe the PR as it is is more consistent with that. OTOH I do see the benefit of fewer options, and less risk for confusion.

In the long term, I think it would be good to remove emscripten exceptions and sjlj support from LLVM entirely. We can add a lowering pass in binaryen that reduces those features to wasm MVP. And emcc would run that pass automatically when the enabled features determine that. If we can get there, then the options situation would get simpler (also for exceptions). So I don't feel strongly either way I guess about the specifics of the options for now.

File: tests/test_core.py [hunk 6]
```diff
@@ -111,6 +111,32 @@ def metafunc(self, native_exceptions):
...
+  def metafunc(self, native_sjlj):
+    if native_sjlj:
+      # Wasm SjLj is currently supported only in Wasm backend and V8
+      if not self.is_wasm():
+        self.skipTest('wasm2js does not support Wasm SjLj')
+      self.require_v8()
+      self.set_setting('SJLJ_HANDLING')
+      # These are for Emscripten EH/SjLj
+      self.set_setting('SUPPORT_LONGJMP', 0)
+      self.set_setting('DISABLE_EXCEPTION_THROWING')
```
Comment (kripken) [inline tests/test_core.py:128]: why do we need to disable exception throwing?

File: tools/building.py [hunk 7]
```diff
@@ -323,10 +323,14 @@ def llvm_backend_args():
     allowed = ','.join(settings.EXCEPTION_CATCHING_ALLOWED)
     args += ['-emscripten-cxx-exceptions-allowed=' + allowed]
 
+  # asm.js-style setjmp/longjmp handling
   if settings.SUPPORT_LONGJMP:
-    # asm.js-style setjmp/longjmp handling
     args += ['-enable-emscripten-sjlj']
 
+  # setjmp/longjmp handling using Wasm EH
+  if settings.SJLJ_HANDLING:
```
Comment (kripken) [inline tools/building.py:331]: this could be an `elif` I think?

File: tools/building.py (hunk 5 above)
Comment (kripken) [reply tools/building.py:468]: I see, thanks!

**Comment by sbc100** [comment]
Yes I think its fine for `SUPPORT_LONGJMP=1` to mean `SUPPORT_LONGJMP=emscripten`.  

The good news is that because `SUPPORT_LONGJMP` is the default almost all the users of this option will be using the `SUPPORT_LONGJMP=0` option.

File: src/settings.js (hunk 2 above)
Comment (aheejin) [reply src/settings.js:1765]: Yeah it is unfortunately not simple. Also for Emscripten exception there are two options: `DISABLE_EXCEPTION_THROWING` and `DISABLE_EXCEPTION_CATCHING`, so having a single option to control all EH is not easy because of that too.

Here I added `SJLJ_HANLDING` to be consistent with `EXCEPTION_HANDLING`, but it is also not easy as I described, because we have to appropriate error out or decide which one to honor when conflicting options are given. I don't have a strong preference here, but at least using a single option `SUPPORT_LONGJMP` to control all longjmp options looks simpler.

By the way, there are many lines like this in js libraries:
```js
#if SUPPORT_LONGJMP
```
if `SUPPORT_LONGJMP` is a string, can we do things like
```js
#if SUPPORT_LONGJMP == 'emscripten'
```
?

Comment (kripken) [reply src/settings.js:1765]: Yes, string comparisons in ifdefs should work.

Comment (sbc100) [reply src/settings.js:1765]: I only see a single use of this setting in the library JS files: https://github.com/emscripten-core/emscripten/blob/1267242815c355d7a0b86a7c667790dc4bdcdcda/src/library.js#L1499

And it looks like its doing `if !SUPPORT_LONGJMP` to test if there is no support at all.. which I think will continue to work, no?

Comment (aheejin) [reply src/settings.js:1765]: @sbc100 There's one more usage of `SUPPORT_LONGJMP` in preamble.js: https://github.com/emscripten-core/emscripten/blob/73c20f60cfa046e156a2c3970f7967b00e6662e7/src/preamble.js#L689

And the convention for multiple-option settings in https://github.com/emscripten-core/emscripten/blob/main/src/settings.js seems to something like `optionA`, `optionB`, and `none`; examples are:  https://github.com/emscripten-core/emscripten/blob/73c20f60cfa046e156a2c3970f7967b00e6662e7/src/settings.js#L1933-L1947
https://github.com/emscripten-core/emscripten/blob/73c20f60cfa046e156a2c3970f7967b00e6662e7/src/settings.js#L118-L138
https://github.com/emscripten-core/emscripten/blob/73c20f60cfa046e156a2c3970f7967b00e6662e7/src/settings.js#L1861-L1867

So I think we can make this option take one of the three values: `wasm`, `emscripten`, and `none`. And For backward compatibility, 1 will be converted to `emscripten` and 0 will be converted to `none` in  emcc.py. That's the reason I asked.. If the string comparisons are fine there will be no problem.

File: tests/test_core.py (hunk 6 above)
Comment (kripken) [reply tests/test_core.py:137]: Plan sounds good. Is it documented somewhere?

File: emcc.py [hunk 8]
```diff
@@ -1369,6 +1369,18 @@ def default_setting(name, new_default):
...
+  if settings.SUPPORT_LONGJMP == 0:
+    settings.SUPPORT_LONGJMP = 'none'
+  if settings.SUPPORT_LONGJMP == 1:
+    settings.SUPPORT_LONGJMP = 'emscripten'
+  # Wasm SjLj cannot be used with Emscripten EH
+  if not settings.DISABLE_EXCEPTION_CATCHING and settings.SUPPORT_LONGJMP == 'wasm':
+    exit_with_error('SUPPORT_LONGJMP=wasm cannot be used with DISABLE_EXCEPTION_CATCHING=0')
+  # DISABLE_EXCEPTION_THROWING=0 is the default, so we change it here
+  if settings.SUPPORT_LONGJMP == 'wasm':
+    settings.DISABLE_EXCEPTION_THROWING = 1
```
Comment (kripken) [inline emcc.py:1382]: This last if surprises me. So for now we don't support wasm EH + wasm Sjlj?

If so should this error instead?

File: tools/system_libs.py [hunk 9]
```diff
@@ -615,6 +615,42 @@ def get_default_variation(cls, **kwargs):
...
+    self.is_wasm = kwargs.pop('is_wasm')
+    super().__init__(**kwargs)
+
+  def get_cflags(self):
+    cflags = super().get_cflags()
+    if self.is_wasm:
+      # DISABLE_EXCEPTION_THROWING=1 is the default, which is for Emscripten
+      # EH/SjLj, so we should reverse it.
+      cflags += ['-s', 'SUPPORT_LONGJMP=wasm',
+                 '-s', 'DISABLE_EXCEPTION_THROWING=1',
```
Comment (kripken) [inline tools/system_libs.py:630]: Same question as above, I guess - not sure why we need to adjust the throwing flag. The comment here also is a little confusing for me: what is reversed?

Comment (kripken) [reply tools/system_libs.py:630]: Oh wait, I think my confusion was that catching is disabled by default but not throwing. But then, why do we need to disable throwing? Is that a temporary workaround for something, or fundamental?

Comment (kripken) [reply tools/system_libs.py:630]: Hmm, yes, maybe it doesn't matter.

Could we just do `-fno-exceptions` here? Or would that interfere with wasm exceptions in the future?

Anyhow, I'm ok with any of the options. Banning all exceptions for now seems least ambiguous though, if we can do that one.

File: emcc.py (hunk 8 above)
Comment (kripken) [reply emcc.py:1382]: I agree it's ok as its experimental. Still I think it's more consistent with other stuff to error on combinations we don't support. That might save confusion later. So I'd suggest
```suggestion
    if settings.DISABLE_EXCEPTION_THROWING == 0:
      exit_with_error('wasm sjlj is not supported with emscripten exceptions (and will not be supported)')
    if settings.EXCEPTION_HANDLING == 0:
      exit_with_error('wasm sjlj is not supported with wasm exceptions yet')
```

File: emcc.py [hunk 10]
```diff
@@ -1369,6 +1369,21 @@ def default_setting(name, new_default):
   if settings.DISABLE_EXCEPTION_THROWING and not settings.DISABLE_EXCEPTION_CATCHING:
     exit_with_error("DISABLE_EXCEPTION_THROWING was set (probably from -fno-exceptions) but is not compatible with enabling exception catching (DISABLE_EXCEPTION_CATCHING=0). If you don't want exceptions, set DISABLE_EXCEPTION_CATCHING to 1; if you do want exceptions, don't link with -fno-exceptions")
 
+  # SUPPORT_LONGJMP=1 means SUPPORT_LONGJMP=emscripten (backward compatibility)
+  if settings.SUPPORT_LONGJMP == 0:
+    settings.SUPPORT_LONGJMP = 'none'
+  if settings.SUPPORT_LONGJMP == 1:
```
Comment (sbc100) [inline emcc.py:1375]: elif

File: emcc.py [hunk 11]
```diff
@@ -2015,9 +2030,9 @@ def include_and_export(name):
       elif settings.LINKABLE:
         diagnostics.warning('experimental', '-s LINKABLE + pthreads is experimental')
 
-      default_setting('SUPPORT_LONGJMP', 0)
-      if settings.SUPPORT_LONGJMP:
-        exit_with_error('SUPPORT_LONGJMP is not compatible with pthreads + dynamic linking')
+      default_setting('SUPPORT_LONGJMP', 'none')
+      if settings.SUPPORT_LONGJMP == 'emscripten':
+        exit_with_error('SUPPORT_LONGJMP=emscripten is not compatible with pthreads + dynamic linking')
```
Comment (sbc100) [inline emcc.py:2035]: Maybe leave these two lines as is?  (Since I imagine `SUPPORT_LONGJMP` of any kind is currently not support .. or at least not tested with pthreads + dynamic linking).

File: emcc.py [hunk 12]
```diff
@@ -2367,8 +2382,11 @@ def get_full_import_name(name):
   if settings.WASMFS:
     settings.LINK_AS_CXX = True
 
-  if settings.RELOCATABLE and settings.EXCEPTION_HANDLING:
-    settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE.append('__cpp_exception')
+  if settings.RELOCATABLE:
```
Comment (sbc100) [inline emcc.py:2385]: Perhaps a comment here.  Something like `export Tag objects which are likely needed by the native code, but which are currently not reported in the metadata of wasm-emscripten-finalize`?

File: src/library.js [hunk 13]
```diff
@@ -1409,21 +1409,21 @@ LibraryManager.library = {
...
   _emscripten_throw_longjmp__sig: 'v',
   _emscripten_throw_longjmp: function() { throw 'longjmp'; },
-#if !SUPPORT_LONGJMP
+#if SUPPORT_LONGJMP == 'none'
 #if !INCLUDE_FULL_LIBRARY
   // These are in order to print helpful error messages when either longjmp of
   // setjmp is used.
   longjmp__deps: [function() {
-    error('longjmp support was disabled (SUPPORT_LONGJMP=0), but it is required by the code (either set SUPPORT_LONGJMP=1, or remove uses of it in the project)');
+    error('longjmp support was disabled (SUPPORT_LONGJMP=none), but it is required by the code (either set SUPPORT_LONGJMP=emscripten, or remove uses of it in the project)');
```
Comment (sbc100) [inline src/library.js:1417]: Do we really want to advertise these new settings.. maybe we can keep advertising the old =0 and =1 settings?  Ideally one day we will change the meaning of `=0` and `=1` so maybe they are still useful.. and maybe we shouldn't consider them legacy, but more like default on/off?

File: tools/system_libs.py [hunk 14]
```diff
@@ -650,7 +686,7 @@ def get_default_variation(cls, **kwargs):
     return super().get_default_variation(is_asan=settings.USE_ASAN, **kwargs)
 
 
-class libcompiler_rt(MTLibrary):
+class libcompiler_rt(MTLibrary, SjLjLibrary):
```
Comment (sbc100) [inline tools/system_libs.py:689]: Maybe add a comment as to why this needs to be  a `SjLjLibrary`?

File: emcc.py (hunk 12 above)
Comment (aheejin) [reply emcc.py:2385]: Done

File: src/library.js (hunk 13 above)
Comment (aheejin) [reply src/library.js:1417]: Then what do you want each of the values to represent eventually? `0` means no support and `1` means Wasm SjLj, and we don't provide Emscripten option?

File: emcc.py (hunk 11 above)
Comment (aheejin) [reply emcc.py:2035]: Is there a reason Wasm SjLj is not supported with pthreads + dynamic linking? Dynamic linking should be no problem because we support it the same way we did for Wasm EH, and pthreads is not tested yet, but it's not like "that's an unsupported combination"..?

File: src/library.js [hunk 15]
```diff
@@ -1408,21 +1408,21 @@ LibraryManager.library = {
 
   _emscripten_throw_longjmp__sig: 'v',
   _emscripten_throw_longjmp: function() { throw 'longjmp'; },
-#if !SUPPORT_LONGJMP
+#if SUPPORT_LONGJMP == 'none'
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/library.js:1411]: Done

File: emcc.py [hunk 16]
```diff
@@ -1369,6 +1369,21 @@ def default_setting(name, new_default):
   if settings.DISABLE_EXCEPTION_THROWING and not settings.DISABLE_EXCEPTION_CATCHING:
     exit_with_error("DISABLE_EXCEPTION_THROWING was set (probably from -fno-exceptions) but is not compatible with enabling exception catching (DISABLE_EXCEPTION_CATCHING=0). If you don't want exceptions, set DISABLE_EXCEPTION_CATCHING to 1; if you do want exceptions, don't link with -fno-exceptions")
 
+  # SUPPORT_LONGJMP=1 means SUPPORT_LONGJMP=emscripten (backward compatibility)
+  if settings.SUPPORT_LONGJMP == 0:
+    settings.SUPPORT_LONGJMP = 'none'
```
[the remark this answers is not in the document]
Comment (aheejin) [reply emcc.py:1374]: Done

File: src/library.js (hunk 13 above)
Comment (aheejin) [reply src/library.js:1416]: Done

File: tests/test_other.py [hunk 17]
```diff
@@ -10380,15 +10380,15 @@ def test_LIBRARY_DEBUG(self):
...
 
   def test_SUPPORT_LONGJMP_object(self):
-    # compile the object *with* support, but link without
-    self.run_process([EMCC, test_file('core/test_longjmp.c'), '-c', '-s', 'SUPPORT_LONGJMP', '-o', 'a.o'])
-    stderr = self.run_process([EMCC, 'a.o', '-s', 'SUPPORT_LONGJMP=0'], stderr=PIPE, check=False).stderr
-    self.assertContained('error: longjmp support was disabled (SUPPORT_LONGJMP=0), but it is required by the code (either set SUPPORT_LONGJMP=1, or remove uses of it in the project)',
+    # compile the object *with* Emscripten SjLj support, but link without
+    self.run_process([EMCC, test_file('core/test_longjmp.c'), '-c', '-s', 'SUPPORT_LONGJMP=emscripten', '-o', 'a.o'])
+    stderr = self.run_process([EMCC, 'a.o', '-s', 'SUPPORT_LONGJMP=none'], stderr=PIPE, check=False).stderr
+    self.assertContained('error: longjmp support was disabled (SUPPORT_LONGJMP=none), but it is required by the code (either set SUPPORT_LONGJMP=emscripten, or remove uses of it in the project)',
```
[the remark this answers is not in the document]
Comment (aheejin) [reply tests/test_other.py:10391]: Done

File: tools/system_libs.py [hunk 18]
```diff
@@ -650,7 +686,7 @@ def get_default_variation(cls, **kwargs):
     return super().get_default_variation(is_asan=settings.USE_ASAN, **kwargs)
 
 
-class libcompiler_rt(MTLibrary):
+class libcompiler_rt(MTLibrary, SjLjLibrary):  # ASan uses setjmp/longjmp
```
[the remark this answers is not in the document]
Comment (aheejin) [reply tools/system_libs.py:689]: Oh didn't know that. Then we should make asan `SjLjLibrary` instead I guess?

[the remark this answers is not in the document]
Comment (aheejin) [reply tools/system_libs.py:689]: Oh no, this is necessary for another reason, because of this part: https://github.com/emscripten-core/emscripten/pull/14976/files#diff-444eb2dc4484af33db2401becfb685bbab1916088c79d969d7c87406d9a39e27R71-R90

This `__USING_WASM_SJLJ__` is enabled when `is_wasm` in `SjLjLibrary` is set. We can't compile this part unconditionally because [this line](https://github.com/emscripten-core/emscripten/pull/14976/files#diff-444eb2dc4484af33db2401becfb685bbab1916088c79d969d7c87406d9a39e27R87) containing a call to `__builtin_wasm_throw` will not compile unless `is_wasm` in `SjLjLibrary` is set, which sets `SUPPORT_LONGJMP=wasm`, which in turn [sets](https://github.com/emscripten-core/emscripten/pull/14976/files#diff-e6615de4338b98e08471c53f2d6fc250dc6ca22cd8c303ac33ab259c18ae68d2R252-R253) `-mllvm -wasm-enable-sjlj`, which enables EH capability in LLVM.

File: tests/test_core.py [hunk 19]
```diff
@@ -1052,14 +1090,15 @@ def test_setjmp_many_2(self):
...
 
+  @with_both_sjlj_handling
   def test_setjmp_noleak(self):
     self.do_runf(test_file('core/test_setjmp_noleak.c'), 'ok.')
 
   @with_both_exception_handling
   def test_exceptions(self):
     self.set_setting('EXCEPTION_DEBUG')
     self.maybe_closure()
-    for support_longjmp in [0, 1]:
```
Comment (sbc100) [inline tests/test_core.py:1062]: Revert all the changes here and below in this file?

File: tools/system_libs.py [hunk 20]
```diff
@@ -615,6 +615,42 @@ def get_default_variation(cls, **kwargs):
...
+    else:
+      cflags += ['-s', 'SUPPORT_LONGJMP=emscripten']
+    return cflags
+
+  def get_base_name(self):
+    name = super().get_base_name()
+    # TODO Currently emscripten-based SjLj is the default mode, thus no
+    # suffixes. Change the default to wasm exception later.
+    if self.is_wasm:
+      name += '-wasm'
```
Comment (sbc100) [inline tools/system_libs.py:641]: libcompiler-rt-wasm.a seems a little ambiguous maybe? 

How about `wasm-exceptions` or `wasm-eh`?

File: tools/system_libs.py [hunk 21]
```diff
@@ -615,6 +615,42 @@ def get_default_variation(cls, **kwargs):
     return super().get_default_variation(eh_mode=eh_mode, **kwargs)
 
 
+class SjLjLibrary(Library):
```
Comment (sbc100) [inline tools/system_libs.py:618]: How about calling this `LongjmpLibrary`?  Would that better describe its purpose?

File: tools/system_libs.py [hunk 22]
```diff
@@ -1370,7 +1407,7 @@ class liblsan_rt(SanitizerLibrary):
                       'lsan_common_emscripten.cpp']
 
 
-class libasan_rt(SanitizerLibrary):
+class libasan_rt(SanitizerLibrary, SjLjLibrary): # ASan uses setjmp/longjmp
```
Comment (sbc100) [inline tools/system_libs.py:1410]: Is it just `libasan` that need this variant and not any of the other sanitizer libraries?  

The only reference I can see to longjmp in asan code seems to be `system/lib/compiler-rt/lib/asan/asan_interceptors.cpp` and that file looks like its ifdef'd out for emscripten.. so is this needed?

File: tests/test_core.py (hunk 19 above)
Comment (aheejin) [reply tests/test_core.py:1062]: Done

File: tools/system_libs.py (hunk 21 above)
Comment (aheejin) [reply tools/system_libs.py:618]: I don't have strong opinions on names, but the functionality handles both setjmp and longjmp, so why would say only longjmp? But yeah we already have the setting name as `SUPPORT_LONGJMP`, so it might be better to be consistent, but all LLVM options use `sjlj`. What's your reason for using 'longjmp' in Emscripten side?

File: tools/system_libs.py (hunk 20 above)
Comment (aheejin) [reply tools/system_libs.py:641]: I think either is fine; Will wait until we decide on the name `SjLjLibrary` vs. `LongjmpLibrary`, and rename this to match that.

File: tools/system_libs.py (hunk 22 above)
Comment (aheejin) [reply tools/system_libs.py:1410]: Oh didn't know it was ifdef'd out. I only grepped files.. Thanks, will remove it.

File: tests/test_core.py [hunk 23]
```diff
@@ -1454,17 +1493,17 @@ def test_iostream_ctors(self):
 ''', 'bugfree code')
 
   def test_exceptions_longjmp1(self):
-    self.set_setting('SUPPORT_LONGJMP')
+    self.set_setting('SUPPORT_LONGJMP', 'emscripten')
```
Comment (sbc100) [inline tests/test_core.py:1496]: Revert this change and the two below?

Comment (aheejin) [reply tests/test_core.py:1496]: Oh I missed them last time. Done.

**Comment by aheejin** [comment]
Thanks for the careful review!
